### PR TITLE
Hold the state

### PR DESCRIPTION
**Pull request by oli-obk**
precursor to #14

ok... iterator style was much easier than anticipated... The current design would require streaming iterators though...

Basically it returns the statement or the terminator that is going to be interpreted *next*. Then it actually interprets that at the next `step` call and returns the statement after that.

### CODE REVIEW

File: src/interpreter/iterator.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    pub(super) fn new(fncx: &'fncx mut FnEvalContext<'a, 'b, 'mir, 'tcx>) -> Self {
+        Stepper {
+            block: fncx.frame().next_block,
+            mir: fncx.mir(),
+            fncx: fncx,
+            stmt: 0,
+            process: Self::dummy,
+        }
+    }
+    fn dummy(&mut self) -> EvalResult<()> { Ok(()) }
```
Comment (solson) [inline src/interpreter/iterator.rs]: Style nit: blank lines between functions

**Comment by solson** [comment]
Do you think it's problematic that the recursive parts (where `call_nested` was called before) aren't part of the iterator, but have their own iterators?

I suppose like @eddyb [said](https://github.com/solson/miri/issues/14#issuecomment-222519492) we could pre-compute statics/constants/promoted-rvalues.

Or on the other hand, in the current design once their resulting allocations are properly cached (and marked immutable in `Memory` somehow) they will effectively be lazily computed, which isn't semantically much different from pre-computing. So I guess it makes sense that they aren't part of the "same execution" (same `Stepper`).

File: src/interpreter/mod.rs [hunk 2]
```diff
@@ -996,13 +982,13 @@ impl<'a, 'b, 'mir, 'tcx> FnEvalContext<'a, 'b, 'mir, 'tcx> {
                 use rustc::mir::repr::Literal::*;
                 match *literal {
                     Value { ref value } => Ok(self.const_to_ptr(value)?),
-                    Item { .. } => unimplemented!(),
+                    Item { .. } => Err(EvalError::Unimplemented(format!("function pointers are unimplemented"))),
```
Comment (solson) [inline src/interpreter/mod.rs]: I believe `Literal::Item` can be more than just function pointers, but I'm not sure. @eddyb?

**Comment by oli-obk** [comment]
We can always add that later. Whenever we enter a function, we could add its promoted values to a Vector which we first process in the `step` function. That messes up the order somewhat, but I think it can be expected that constants are computed before everything else.

For global constants/statics we'd need something else, because that would have to happen during the creation of the global context and before the first function call... Maybe some kind of builder?

> and marked immutable in Memory somehow

not all statics should be immutable.

**Comment by solson** [comment]
> not all statics should be immutable.

@oli-obk In the constant evaluator, according to @eddyb's model, they should be. I believe even `static mut` mutation was to be banned at compile-time.

File: src/interpreter/mod.rs [hunk 3]
```diff
@@ -20,6 +20,8 @@ use error::{EvalError, EvalResult};
 use memory::{Memory, Pointer};
 use primval::{self, PrimVal};
 
+mod iterator;
```
Comment (solson) [inline src/interpreter/mod.rs]: This module name/filename is confusing given that it has no `Iterator` impl. `mod stepper` perhaps?

File: src/interpreter/mod.rs (hunk 2 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: yea it can be more than that... but most of it is tagged with `FIXME`s

Comment (solson) [inline src/interpreter/mod.rs]: I'd be happy with `"literal items (e.g. function pointers) are unimplemented"` just to avoid the possibility of future confusion.

File: src/interpreter/mod.rs (hunk 3 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: done

File: src/interpreter/mod.rs (hunk 2 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: done

File: tests/compile-fail/unimplemented.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#![allow(dead_code, unused_attributes)]
+
+//error-pattern:literal items (e.g. mentions of function items) are unimplemented
+
+static mut X: usize = 5;
+
+#[miri_run]
+fn static_mut() {
+    unsafe {
+        X = 6;
```
Comment (solson) [inline tests/compile-fail/unimplemented.rs]: So yeah, this might fail for a different reason in the future (mutation of a static), but I'm fine with having it here so we notice when that happens.

Although it would be allowed in versions of Miri not intended for rustc const eval.

**Comment by eddyb** [comment]
LGTM, modulo the lack of constant pre-computation. Do note that when part of this gets merged into rustc, some of the constants will already have been computed by rustc itself and may have caused compiler errors, possibly before miri can really take over.

**Comment by oli-obk** [comment]
> some of the constants will already have been computed by rustc itself

Why only some?

> may have caused compiler errors, possibly before miri can really take over.

Interpreting code with compiler errors is probably not useful anyway. Miri [already aborts if there were compiler errors](https://github.com/solson/miri/blob/master/src/bin/miri.rs#L27).

> the lack of constant pre-computation

It's easy to implement for promoted constants. But there's no list of global constants/statics...

Also evaluating a constant might require evaluating another constant. So even if we cache constants, we'd need to evaluate them in an order that makes sure that all required constants are computed before the requiring constant.

I'm not sure how to proceed...

**Comment by eddyb** [comment]
> Why only some?

Because you may not be able to evaluate, e.g. an associated constant used from a function which is also generic - only trans-time monomorphization (or miri, in this case) will trigger its evaluation.

> It's easy to implement for promoted constants. But there's no list of global constants/statics...

Not really, what I meant is collecting all *uses* of constants in a function, and having that as a different set of states, before any statement/terminator.
When finishing evaluation of a constant, you pop a "stack frame" and go to the next pending constant in the "caller", and if there aren't any left, you go to `(START_BLOCK, 0)`.

**Comment by oli-obk** [comment]
I refactored the code to cache statics globally and cache promoteds just for the current stackframe. This has the unfortunate side-effect, that statics can now be written to.

I also removed the current recursion that happens when evaluating a static/constant, and now the stepper simply pushes another stack frame for constants/statics

**Comment by oli-obk** [comment]
The way it's written now, we technically don't need to do the pre-evaluation of promoteds and constants... we could just do it on-demand.

**Comment by eddyb** [comment]
@oli-obk How can you easily do it on-demand? You can have arbitrarily many constant operands in one statement/terminator, even the simplest ones.
This is in part because of the indexing on lvalue, i.e. `var0[A][B][C][D]` is a valid `Lvalue`.

**Comment by eddyb** [comment]
@oli-obk Unless you mean just before the statement/terminator, which I guess is a fine approach.

**Comment by oli-obk** [comment]
> This is in part because of the indexing on lvalue, i.e. var0[A][B][C][D] is a valid Lvalue.

oh right, so we wouldn't save the allocation of the vector... let's leave it as is.

**Comment by oli-obk** [comment]
some benchmarks, not much of a change compared with the [last ones](https://github.com/solson/miri/pull/13#issue-157490823)

```
test fib           ... bench:         111 ns/iter (+/- 3)
test fib_iter      ... bench:           1 ns/iter (+/- 0)
test fib_iter_miri ... bench:   8,813,523 ns/iter (+/- 296,007)
test fib_miri      ... bench:   5,816,824 ns/iter (+/- 259,427)
```

**Comment by oli-obk** [comment]
this PR ended up being able to evaluate fewer functions than before, because there might be statics/promoteds that aren't needed in the interpretation and thus weren't interpreted before. If those statics/constants contained code that miri can't interpret (yet), the entire function evaluation failed. I added a refactoring of the call stack and moved to interpret constants only just before the statement/terminator that needs them.

**Comment by solson** [comment]
Sorry for letting this sit. I'll be reviewing it tomorrow.

File: src/interpreter/stepper.rs [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    Constant,
+    Assignment,
+    Terminator,
+    Done,
+}
+
+pub struct Stepper<'fncx, 'a: 'fncx, 'b: 'a + 'mir, 'mir: 'fncx, 'tcx: 'b>{
+    fncx: &'fncx mut FnEvalContext<'a, 'b, 'mir, 'tcx>,
+    mir: CachedMir<'mir, 'tcx>,
+    process: fn (&mut Stepper<'fncx, 'a, 'b, 'mir, 'tcx>) -> EvalResult<()>,
```
Comment (solson) [inline src/interpreter/stepper.rs]: Can we replace the function pointers with an `enum`? Instead of calling `(self.process)(self)`, match over the enum and call whatever method is needed.

I'm not sure why we'd use a function pointer in the first place.

File: src/interpreter/stepper.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+pub enum Event {
+    Constant,
+    Assignment,
+    Terminator,
+    Done,
+}
+
+pub struct Stepper<'fncx, 'a: 'fncx, 'b: 'a + 'mir, 'mir: 'fncx, 'tcx: 'b>{
+    fncx: &'fncx mut FnEvalContext<'a, 'b, 'mir, 'tcx>,
+    mir: CachedMir<'mir, 'tcx>,
```
Comment (solson) [inline src/interpreter/stepper.rs]: Can we remove this field and just use `fncx.mir()` all the time?

File: src/interpreter/frame.rs [hunk 7]
```diff
@@ -0,0 +1,923 @@
...
+    pub temp_offset: usize,
+
+    /// The index of the currently evaluated statment
+    pub stmt: usize,
+
+    // Constants that need to be evaluated before the next statement can be evaluated
+    pub constants: Vec<(ConstantId<'tcx>, Span, Pointer, CachedMir<'mir, 'tcx>)>,
+}
+
+pub(super) struct FrameInterpreter<'a, 'b: 'a, 'mir: 'a, 'tcx: 'a + 'mir + 'b> {
```
Comment (solson) [inline src/interpreter/frame.rs]: I don't understand the `FrameInterpreter`/`FnEvalContext` split. Why is statement evaluation here and terminator evaluation over there?

I want to lean towards fewer "context" types if possible. I wasn't really happy I had to make the `Global`/`Fn` split, either.

File: src/interpreter/frame.rs [hunk 8]
```diff
@@ -0,0 +1,923 @@
...
+        self.gecx.monomorphize(ty, self.frame.substs)
+    }
+
+    pub fn type_size(&self, ty: Ty<'tcx>) -> usize {
+        self.gecx.type_size(ty, self.frame.substs)
+    }
+
+    pub fn type_layout(&self, ty: Ty<'tcx>) -> &'tcx Layout {
+        self.gecx.type_layout(ty, self.frame.substs)
+    }
```
Comment (solson) [inline src/interpreter/frame.rs]: Forwarding functions like the above three are why I made `FnEvalContext` deref to the `GlobalEvalContext` (like how `TyCtxt` works in rustc).

File: src/interpreter/global.rs [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        let substituted = ty.subst(self.tcx, substs);
+        self.tcx.normalize_associated_type(&substituted)
+    }
+
+    pub fn type_size(&self, ty: Ty<'tcx>, substs: &'tcx Substs<'tcx>) -> usize {
+        self.type_layout(ty, substs).size(&self.tcx.data_layout).bytes() as usize
+    }
+
+    pub fn type_layout(&self, ty: Ty<'tcx>, substs: &'tcx Substs<'tcx>) -> &'tcx Layout {
+        if let Some(layout) = self.monomorphizations.borrow().get(&(ty, substs)) {
```
Comment (solson) [inline src/interpreter/global.rs]: `monomorphizations` should cache the results of `monomorphize`, no? Sometimes we need the monomorphized type itself, and getting that should be fast, too.

Caching the `Ty` -> `Layout` steps could be done separately.

File: src/interpreter/global.rs [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+            return layout;
+        }
+        debug!("monomorphising: {:?} + {:?}", ty, substs);
+        // TODO(solson): Is this inefficient? Needs investigation.
```
Comment (solson) [inline src/interpreter/global.rs]: I think we can remove this comment now.

File: src/interpreter/mod.rs [hunk 11]
```diff
@@ -37,17 +37,6 @@ struct GlobalEvalContext<'a, 'tcx: 'a> {
...
-    /// exists separately from `stack` because it must contain the `Substs` for a function while
-    /// *creating* the `Frame` for that same function.
-    substs_stack: Vec<&'tcx Substs<'tcx>>,
-
-    // TODO(solson): Merge with `substs_stack`. Also try restructuring `Frame` to accomodate.
-    /// A stack of the things necessary to print good strack traces:
-    ///   * Function DefIds and Substs to print proper substituted function names.
-    ///   * Spans pointing to specific function calls in the source.
-    name_stack: Vec<(DefId, &'tcx Substs<'tcx>, codemap::Span)>,
-
```
Comment (solson) [inline src/interpreter/mod.rs]: Hmm, I guess cleaning up the stacks was a lot easier than I thought (particularly `substs_stack`).

We definitely need a ton more tests dealing with generic functions, monomorphization, and substs to make sure we haven't screwed it up, but your change here seems correct to me.

Thanks for the cleanups. \o/

File: src/interpreter/mod.rs [hunk 12]
```diff
@@ -129,9 +129,9 @@ enum TerminatorTarget {
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
-enum ConstantId {
+enum ConstantId<'tcx> {
     Promoted { index: usize },
-    Static { def_id: DefId },
+    Static { def_id: DefId, substs: &'tcx Substs<'tcx> },
```
Comment (solson) [inline src/interpreter/mod.rs]: Can `Promoted` constants requires substs, too? For example:

```rust
struct Foo<T>(T);
fn foo() {
    let x = &Foo(42); // The Foo(42) constant could be promoted, right?
}
```

@eddyb ^ confirm/deny?

File: src/interpreter/mod.rs [hunk 13]
```diff
@@ -178,6 +172,16 @@ impl<'a, 'b, 'mir, 'tcx> FnEvalContext<'a, 'b, 'mir, 'tcx> {
         }
     }
 
+    fn alloc_ret_ptr(&mut self, ty: ty::FnOutput<'tcx>) -> Option<Pointer> {
```
Comment (solson) [inline src/interpreter/mod.rs]: Nit: rename the `ty` argument to avoid shadowing below? Maybe `output_ty`.

File: src/interpreter/stepper.rs [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx> {
+    fn visit_constant(&mut self, constant: &mir::Constant<'tcx>) {
+        self.super_constant(constant);
+        match constant.literal {
+            // already computed by rustc
+            mir::Literal::Value { .. } => {}
+            mir::Literal::Item { def_id, substs } => {
+                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
+                if item_ty.ty.is_fn() {
+                    // unimplemented
```
Comment (solson) [inline src/interpreter/stepper.rs]: `unimplemented!()`

File: src/interpreter/stepper.rs [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx> {
+    fn visit_constant(&mut self, constant: &mir::Constant<'tcx>) {
+        self.super_constant(constant);
+        match constant.literal {
+            // already computed by rustc
+            mir::Literal::Value { .. } => {}
+            mir::Literal::Item { def_id, substs } => {
+                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
```
Comment (solson) [inline src/interpreter/stepper.rs]: Mmm, I believe `subst` alone is not enough in the case that the item type is an associated type. It should use the `monomorphize` method (on `gecx`, IIRC) which also does associated type normalization.

File: src/interpreter/stepper.rs [hunk 16]
```diff
@@ -163,37 +179,64 @@ impl<'fncx, 'a, 'b: 'a + 'mir, 'mir, 'tcx: 'b> Stepper<'fncx, 'a, 'b, 'mir, 'tcx
...
+                } else {
+                    self.constant(def_id, substs, constant.span);
+                }
+            },
             mir::Literal::Promoted { index } => {
-                self.constants.push((ConstantId::Promoted { index: index }, constant.span));
+                if self.fncx.frame().promoted.contains_key(&index) {
+                    return;
+                }
+                let mir = self.mir.promoted[index].clone();
```
Comment (solson) [inline src/interpreter/stepper.rs]: This is cloning an entire `Mir` structure... we should try to find a way around that.

File: src/interpreter/stepper.rs [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        }
+    }
+
+    /// returns the statement that will be processed next
+    pub fn stmt(&self) -> &mir::Statement {
+        &self.fncx.basic_block().statements[self.fncx.frame().stmt]
+    }
+
+    /// returns the terminator of the current block
+    pub fn term(&self) -> &mir::Terminator {
```
Comment (solson) [inline src/interpreter/stepper.rs]: Not a great abbreviation, IMO. s/term/terminator/

File: src/interpreter/mod.rs [hunk 18]
```diff
@@ -118,8 +118,8 @@ enum CachedMir<'mir, 'tcx: 'mir> {
 
 /// Represents the action to be taken in the main loop as a result of executing a terminator.
 enum TerminatorTarget {
-    /// Make a local jump to the given block.
-    Block(mir::BasicBlock),
+    /// Make a local jump to the next block
+    Block,
```
Comment (solson) [inline src/interpreter/mod.rs]: I don't understand the purpose of this commit.

File: src/interpreter/mod.rs (hunk 12 above)
Comment (solson) [inline src/interpreter/mod.rs]: Nevermind, this is fixed in a later commit.

File: src/interpreter/mod.rs [hunk 19]
```diff
@@ -136,10 +133,28 @@ enum TerminatorTarget {
 
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
 enum ConstantId<'tcx> {
```
Comment (solson) [inline src/interpreter/mod.rs]: Perhaps make a `struct` that always has `DefId` and `Substs` and avoid the need for the methods.

**Comment by solson** [comment]
Alright, I reviewed nearly all the code. Some of my comments are marked as "on an outdated diff" by GitHub but I think they're all still relevant, aside from one I noticed was already fixed later.

My main areas of concern are the `Stepper` implementation and the new `FrameInterpreter` struct, and I left some comments asking about them.

File: src/interpreter/mod.rs (hunk 12 above)
Comment (eddyb) [inline src/interpreter/mod.rs]: Why would statics though?

Comment (solson) [inline src/interpreter/mod.rs]: Good point. Statics can't have substs since they aren't allowed to use the function-level type parameters.

File: src/interpreter/stepper.rs (hunk 14 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: this is not necessarily executed, so if we error here, we can't evaluate assertions anymore, because they *may* use a function.

File: src/interpreter/stepper.rs (hunk 15 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: when I do that, rustc panics about 

> error: internal compiler error: ../src/librustc/infer/mod.rs:650: Encountered errors `[FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<T as std::marker::Sized>)),depth=1),Unimplemented)]`
 resolving bounds after type-checking

File: src/interpreter/mod.rs (hunk 18 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: Before it was that when going to a new block, it was the duty of the `eval_terminator` caller to actually change the block. But if `eval_terminator` did a function call, then `eval_terminator` updated the block (see here: https://github.com/solson/miri/pull/18/files/cc1ca73f5736a4aaa7a1ae164788051b8366fa3d..f995db9ffba925590e4c78917362998be02fcbc1#diff-b6ba58183c703f8ffcaa54879146b1c2R326)

File: src/interpreter/mod.rs (hunk 19 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: sounds good

File: src/interpreter/stepper.rs (hunk 5 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: > I'm not sure why we'd use a function pointer in the first place.

because ... reasons ... no idea.

I did a refactoring to get rid of the `Constant` state (I simply push all the constants' stackframes directly, instead of pushing one, evaluating it, pushing the next one, evaluating it, ...)

File: src/interpreter/stepper.rs (hunk 6 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: done

File: src/interpreter/frame.rs (hunk 7 above)
Comment (oli-obk) [inline src/interpreter/frame.rs]: I reverted it.

File: src/interpreter/frame.rs (hunk 8 above)
Comment (oli-obk) [inline src/interpreter/frame.rs]: Yea, I tried to move as much as possible to the `GlobalEventContext`.

File: src/interpreter/frame.rs (hunk 7 above)
Comment (oli-obk) [inline src/interpreter/frame.rs]: And also undid the splitting into files, we can do that in a separate PR

File: src/interpreter/global.rs (hunk 9 above)
Comment (oli-obk) [inline src/interpreter/global.rs]: I undid all the caching, because looking at the traces it's not really clear what's going on. Often times the "same" (textual output is the same) type + subst combo get hashed differently... I don't know enough about this do to it properly

File: src/interpreter/mod.rs (hunk 13 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: done

File: src/interpreter/stepper.rs (hunk 17 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: I removed the method entirely in a refactoring

File: src/interpreter/mod.rs (hunk 19 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: done

**Comment by oli-obk** [comment]
I addressed all the comments. I modified the commits in place, because some refactorings were easier done by simply removing commits (e.g. the function/frame split)

**Comment by oli-obk** [comment]
The major change is the way the stepper works. You pass a closure to the `step` function instead of processing the return value. This makes the code much more concise and prevents us from reloading the entire state at every call. I'm benchmarking currently to figure out if that undoes the performance regression from the constant-jit changes.

**Comment by oli-obk** [comment]
how is travis passing if there's a failure? Can we force travis to work with a specific rustc version?

**Comment by solson** [comment]
Why do the `Stepper` functions take closures? I don't understand the overall goal with that.

File: src/interpreter/global.rs (hunk 9 above)
Comment (solson) [inline src/interpreter/global.rs]: @eddyb ^ Any idea what could make `Substs` hashing go wrong?

File: src/interpreter/stepper.rs (hunk 14 above)
Comment (solson) [inline src/interpreter/stepper.rs]: Oh, I didn't read this code carefully enough before.

Should we *ever* visit these? There's no pre-computation to be done for fn items as far as I can tell, so maybe just replace "unimplemented" with a comment about that.

File: src/interpreter/stepper.rs (hunk 15 above)
Comment (solson) [inline src/interpreter/stepper.rs]: Weird. Can you add `// FIXME: Should probably use self.gecx.monomorphize; currently hits an ICE` so I can find it later?

File: src/interpreter/mod.rs (hunk 18 above)
Comment (solson) [inline src/interpreter/mod.rs]: So you were just making `Block` more similar to `Call`? It's not actually required by the rest of the PR?

Just wondering. I don't mind the PR, and I'll probably look for other ways to refactor `TerminatorTarget`.

**Comment by solson** [comment]
Aside from the closures, `Stepper` makes more sense to me than before.

My main concern now is the `Mir` cloning I mentioned in https://github.com/solson/miri/pull/18#discussion_r66188751.

File: src/interpreter/mod.rs (hunk 18 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: Well... it caused bugs in the Stepper because I had to very carefully figure out where to get the next block from. Now that I use `Frame::next_block` this might be way less relevant...

I think we might be able to get rid of `TerminatorTarget` completely

Comment (solson) [inline src/interpreter/mod.rs]: > I think we might be able to get rid of TerminatorTarget completely
 
That was my thought. We can merge this PR as-is though.

File: src/interpreter/stepper.rs (hunk 15 above)
Comment (eddyb) [inline src/interpreter/stepper.rs]: Was that from `gecx.monomorphize(constant.ty)`? Cause if it was just the result of `lookup_item_type`, that's not enough.
Generally speaking, `constant.ty == self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs)` but the latter is more effort.

Comment (oli-obk) [inline src/interpreter/stepper.rs]: face -> palm ... I'm fixing this

**Comment by oli-obk** [comment]
> What if we inlined the trace! calls into Stepper and only had to worry about done? Could it get simplified down and not require the closures?

done.

> My main concern now is the Mir cloning I mentioned in #18 (comment).

We can cache it to ensure that the cloning only ever happens once for every promoted.

File: src/interpreter/stepper.rs [hunk 20]
```diff
@@ -151,9 +151,10 @@ impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx>
             // already computed by rustc
             mir::Literal::Value { .. } => {}
             mir::Literal::Item { def_id, substs } => {
-                let item_ty = self.gecx.tcx.lookup_item_type(def_id).subst(self.gecx.tcx, substs);
```
Comment (solson) [inline src/interpreter/stepper.rs]: I think this change also applies to https://github.com/solson/miri/pull/18/files#diff-b6ba58183c703f8ffcaa54879146b1c2R1192

File: src/interpreter/stepper.rs [hunk 21]
```diff
@@ -151,9 +151,10 @@ impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx>
...
-                if item_ty.ty.is_fn() {
-                    // unimplemented
+                if constant.ty.is_fn() {
```
Comment (eddyb) [inline src/interpreter/stepper.rs]: Does this only return true for `TyFnDef`?

File: src/interpreter/stepper.rs [hunk 22]
```diff
@@ -151,9 +151,10 @@ impl<'a, 'b, 'mir, 'tcx> Visitor<'tcx> for ConstantExtractor<'a, 'b, 'mir, 'tcx>
...
+                    // No need to do anything here, even if function pointers are implemented,
+                    // because the type is the actual function, not the signature of the function.
+                    // Thus we can simply create a zero sized allocation in `evaluate_operand`
                 } else {
                     self.static_item(def_id, substs, constant.span);
```
Comment (eddyb) [inline src/interpreter/stepper.rs]: Should probably be named `global_item` or some such.

File: src/interpreter/stepper.rs (hunk 20 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: it does, fixed

File: src/interpreter/stepper.rs (hunk 21 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: good catch, fixed

File: src/interpreter/stepper.rs (hunk 22 above)
Comment (oli-obk) [inline src/interpreter/stepper.rs]: done

**Comment by solson** [comment]
I read through the changes one last time. I think I'm ready to merge.

@eddyb Did you have any outstanding comments?

@oli-obk Was there anything you didn't get around to yet that you wanted to?

**Comment by solson** [comment]
I heard from eddyb on IRC, so I'll just merge now.
